### DISCUSSION TITLE
Idea: Deterministic polling

### DISCUSSION THREAD

**Discussion by dot-bob** [darthcloud/BlueRetro#89] [Ideas]
I am not very familiar how the current bluetooth implementation operates with the controllers and the ESP32 of if the idea below if feasible.  If it is like USB where the host polls a device for updates I wonder if it would be possible to poll or request bluetooth updates in sync with console update request.

So in the case of a SNES, bluetooth polling requests would go out every 16.66ms (or a integer multiple) based off of the strobe.  The request then could be shifted by an offset to have it so the request returns with the result right before the strobe is sampled.  Statistics of response time of each connected controller would need to be kept track of to determine the best offset to use each controller.  These offsets would continually be tweaked tweaked to minimize any bluetooth update latency.  If we could dump the statistics it would also let the user know the latency a specific controller has.

**Comment by darthcloud** [comment]
There is no polling, the controllers send interrupt at their convenience. Completely out of our control.

Wii controller do support polling mode but the back and forth increase latency a lot. 

Btw you might be interested in this post:
https://hackaday.io/project/170365-blueretro/log/187443-2020-12-26-update-latency-tests-release-v010
